Output of `postpublish` script missing in both success and failure cases

<!-- Please do your best to fill out all of the sections below! -->

## Current Behavior

<!-- What is the behavior that currently you experience? -->

### When the `postpublish` script exits non-zero

```
> @our-scope/our-package@0.1.0-alpha.156 postpublish /home/runner/work/.../our-package
> ./post-publish-command.sh

lerna ERR! lifecycle "postpublish" errored in "@our-scope/our-package", exiting 1
lerna WARN notice Package failed to publish: @our-scope/our-package
lerna ERR! 1 command failed
lerna ERR! errno "undefined" is not a valid exit code - exiting with code 1
Error: Process completed with exit code 1.
```

### When the `postpublish` script runs successfully

```
lerna info lifecycle @our-scope/our-package@0.1.0-alpha.159~postpublish: @our-scope/our-package@0.1.0-alpha.159

> @our-scope/our-package@0.1.0-alpha.159 postpublish /home/runner/work/.../our-package
> ./post-publish-command.sh

> @our-scope/our-package@0.1.0-alpha.159 ./post-publish-command.sh
> ./post-publish-command.sh
lerna success published @our-scope/our-package 0.1.0-alpha.159
```

## Expected Behavior

<!-- What is the behavior that you expect to happen? -->
<!-- Is this a regression? .i.e Did this used to be the behavior at one point?  -->

### When the `postpublish` script exits non-zero

```
> @our-scope/our-package@0.1.0-alpha.156 postpublish /home/runner/work/.../our-package
> ./post-publish-command.sh

#########################################
###                                   ###
### Output of post-publish-command.sh ###
###                                   ###
#########################################
npm error Lifecycle script `postpublish` failed with error:
npm error code 1
lerna ERR! lifecycle "postpublish" errored in "@our-scope/our-package", exiting 1
lerna WARN notice Package failed to publish: @our-scope/our-package
lerna ERR! 1 command failed
Error: Process completed with exit code 1.
```

### When the `postpublish` script runs successfully

```
lerna info lifecycle @our-scope/our-package@0.1.0-alpha.159~postpublish: @our-scope/our-package@0.1.0-alpha.159

> @our-scope/our-package@0.1.0-alpha.159 postpublish /home/runner/work/.../our-package
> ./post-publish-command.sh

> @our-scope/our-package@0.1.0-alpha.159 ./post-publish-command.sh
> ./post-publish-command.sh
#########################################
###                                   ###
### Output of post-publish-command.sh ###
###                                   ###
#########################################
lerna success published @our-scope/our-package 0.1.0-alpha.159
```


## Steps to Reproduce

<!-- Help us help you by making it easy for us to reproduce your issue! -->

<!-- Can you reproduce this on https://github.com/lerna/repro? -->
<!-- If so, open a PR with your changes and link it below. -->
<!-- If not, please provide a minimal Github repo of your own -->
<!-- At the very least, provide as much detail as possible to help us reproduce the issue -->

1. In `package.json` have a post-publish script that emits output. Change it to exit 1 or 0
   depending on which behavior you want to reproduce.

2. Publish the package with `lerna publish from-package --dist-tag develop --yes`.

`package.json`
```json
  "scripts": {
    "postpublish": "./post-publish-commmand.sh",
  }
```

`post-publish-command.sh` (make sure it's executable)
```bash
#!/usr/bin/env bash
cat <<EOF
#########################################
###                                   ###
### Output of post-publish-command.sh ###
###                                   ###
#########################################
EOF
exit 1
```

---

By moving the script to `prepublishOnly` (as a sanity check), the output is present. After fixing the issue (in my case a missing required environment variable), I move the script back to `postpublish` and things succeeded, however still the output from the script is being swallowed.

* **Note:** Nothing was done to test `stderr`, however it may also be swallowed.

* Running `npm run postpublish` works fine.

* Running `npm publish` also works fine.

### Failure Logs / Configuration

<!-- Please include any relevant log snippets or files here. -->

<details><summary>lerna.json</summary><p>
<!-- browsers demand the next line be empty -->

```json
{
  "$schema": "node_modules/lerna/schemas/lerna-schema.json",
  "version": "independent",
  "command": {
    "version": {
      "allowBranch": ["main", "develop", "hotfix/*"]
    }
  }
}
```

</p></details>

### Environment

<!-- It's important for us to know the context in which you experience this behavior! -->
<!-- Please paste the result of running `npx lerna info` in your repo below! -->

* Running under GitHub Actions.

* Invoked with `lerna publish from-package --dist-tag develop --yes`.

```
lerna notice cli v8.1.8
lerna info versioning independent
lerna info ci enabled
 Environment info:
  System:
    OS: Linux 6.5 Ubuntu 22.04.5 LTS 22.04.5 LTS (Jammy Jellyfish)
    CPU: (2) x64 AMD EPYC 7763 64-Core Processor
  Binaries:
    Node: 20.17.0 - /opt/hostedtoolcache/node/20.17.0/x64/bin/node
    Yarn: 1.22.22 - /usr/local/bin/yarn
    npm: 10.8.2 - /opt/hostedtoolcache/node/20.17.0/x64/bin/npm
  Utilities:
    Git: 2.46.2 - /usr/bin/git
```

--- a/e2e/publish/src/publish-lifecycle-output.spec.ts
+++ b/e2e/publish/src/publish-lifecycle-output.spec.ts
@@ -1,0 +1,119 @@
+import { Fixture, normalizeCommitSHAs, normalizeEnvironment, trimEnds } from "@lerna/e2e-utils";
+
+const randomInt = (min: number, max: number) => Math.floor(Math.random() * (max - min + 1)) + min;
+const randomVersion = () => `${randomInt(10, 89)}.${randomInt(10, 89)}.${randomInt(10, 89)}`;
+
+expect.addSnapshotSerializer({
+  serialize(str: string) {
+    return trimEnds(
+      normalizeCommitSHAs(normalizeEnvironment(str))
+        .replaceAll(/integrity:\s*.*/g, "integrity: XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX")
+        .replaceAll(/\d*B package\.json/g, "XXXB package.json")
+        .replaceAll(/size:\s*\d*\s?B/g, "size: XXXB")
+        .replaceAll(/\d*\.\d*\s?kB/g, "XXX.XXX kb")
+    );
+  },
+  test(val: string) {
+    return val != null && typeof val === "string";
+  },
+});
+
+describe("lerna-publish-lifecycle-output", () => {
+  let fixture: Fixture;
+
+  beforeEach(async () => {
+    fixture = await Fixture.create({
+      e2eRoot: process.env.E2E_ROOT,
+      name: "lerna-publish",
+      packageManager: "npm",
+      initializeGit: true,
+      lernaInit: { args: [`--packages="packages/*"`] },
+      installDependencies: true,
+    });
+  });
+  afterEach(() => fixture.destroy());
+
+  // Regression test for https://github.com/lerna/lerna/issues/4090
+  it("should show postpublish lifecycle script output", async () => {
+    await fixture.lerna("create test-1 -y");
+
+    await fixture.addScriptsToPackage({
+      packagePath: "packages/test-1",
+      scripts: {
+        postpublish: "echo POSTPUBLISH_OUTPUT_TEST_1",
+      },
+    });
+
+    await fixture.createInitialGitCommit();
+    await fixture.exec("git push origin test-main");
+
+    const version = randomVersion();
+    await fixture.lerna(`version ${version} -y`);
+
+    await fixture.exec('echo "registry=http://localhost:4873" > .npmrc');
+
+    const output = await fixture.lerna("publish from-git -y");
+    await fixture.exec(`npm unpublish --force test-1@${version} --registry=http://localhost:4873`);
+
+    expect(output.combinedOutput).toContain("POSTPUBLISH_OUTPUT_TEST_1");
+  });
+
+  // Regression test for https://github.com/lerna/lerna/issues/4090
+  it("should show publish lifecycle script output", async () => {
+    await fixture.lerna("create test-1 -y");
+
+    await fixture.addScriptsToPackage({
+      packagePath: "packages/test-1",
+      scripts: {
+        publish: "echo PUBLISH_OUTPUT_TEST_1",
+      },
+    });
+
+    await fixture.createInitialGitCommit();
+    await fixture.exec("git push origin test-main");
+
+    const version = randomVersion();
+    await fixture.lerna(`version ${version} -y`);
+
+    await fixture.exec('echo "registry=http://localhost:4873" > .npmrc');
+
+    const output = await fixture.lerna("publish from-git -y");
+    await fixture.exec(`npm unpublish --force test-1@${version} --registry=http://localhost:4873`);
+
+    expect(output.combinedOutput).toContain("PUBLISH_OUTPUT_TEST_1");
+  });
+
+  // Regression test for https://github.com/lerna/lerna/issues/4090
+  it("should show postpublish output from multiple packages", async () => {
+    await fixture.lerna("create test-1 -y");
+    await fixture.lerna("create test-2 -y");
+
+    await fixture.addScriptsToPackage({
+      packagePath: "packages/test-1",
+      scripts: {
+        postpublish: "echo POSTPUBLISH_PKG1",
+      },
+    });
+    await fixture.addScriptsToPackage({
+      packagePath: "packages/test-2",
+      scripts: {
+        postpublish: "echo POSTPUBLISH_PKG2",
+      },
+    });
+
+    await fixture.createInitialGitCommit();
+    await fixture.exec("git push origin test-main");
+
+    const version = randomVersion();
+    await fixture.lerna(`version ${version} -y`);
+
+    await fixture.exec('echo "registry=http://localhost:4873" > .npmrc');
+
+    const output = await fixture.lerna("publish from-git -y");
+    await fixture.exec(`npm unpublish --force test-1@${version} --registry=http://localhost:4873`);
+    await fixture.exec(`npm unpublish --force test-2@${version} --registry=http://localhost:4873`);
+
+    expect(output.combinedOutput).toContain("POSTPUBLISH_PKG1");
+    expect(output.combinedOutput).toContain("POSTPUBLISH_PKG2");
+  });
+});

--- a/libs/core/src/lib/npm-publish.spec.ts
+++ b/libs/core/src/lib/npm-publish.spec.ts
@@ -214,9 +214,10 @@ describe("npm-publish", () => {
     }
   );
 
-  it("calls publish lifecycles", async () => {
+  it("calls publish lifecycles with stdio inherit", async () => {
     const options = expect.objectContaining({
       projectScope: "@scope",
+      stdio: "inherit",
     });
 
     await npmPublish(pkg, tarFilePath, {}, conf);

--- a/libs/core/src/lib/npm-publish.ts
+++ b/libs/core/src/lib/npm-publish.ts
@@ -104,6 +104,13 @@ export async function npmPublish(
     result = await otplease((innerOpts) => publish(manifestContent, tarData, innerOpts), opts, otpCache);
   }
 
+  // ensure lifecycle script output is visible on the terminal,
+  // matching the behavior of prepublishOnly in pack-directory.ts
+  // TODO: refactor based on TS feedback
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore
+  opts.stdio = "inherit";
+
   await runLifecycle(pkg, "publish", opts);
   await runLifecycle(pkg, "postpublish", opts);
 

